RecoTracker/LSTCore/src/alpaka/Triplet.h

```diff
@@ -206,63 +206,307 @@ namespace lst {
+    float dPhi = alpaka::math::acos(acc, (AO * AO + BO * BO - AB2) / (2 * AO * BO)); //Law of Cosines
+    float ds = circleRadius / 100 * dPhi;
+    float pz = dz / ds * pt; 
+
+    float p = alpaka::math::sqrt(acc, px * px + py * py + pz * pz);
+    float a = -2.f * k2Rinv1GeVf * 100 * charge;
+    float rou = a / p;
+
+    float rzChiSquared = 0;
+    float error = 0;
+
+    //check the tilted module, side: PosZ, NegZ, Center(for not tilted)
+    float drdz = alpaka::math::abs(acc, modulesInGPU.drdzs[outerOuterLowerModuleIndex]);
+    short side = modulesInGPU.sides[outerOuterLowerModuleIndex];
+    short subdets = modulesInGPU.subdets[outerOuterLowerModuleIndex];
+
+    //calculate residual
+    if (layer3 <= 6 && ((side == lst::Center) or (drdz < 1))) { // for barrel
+      float paraA = r_init * r_init + 2 * (px * px + py * py) / (a * a) + 2 * (y_init * px - x_init * py) / a - r3 * r3;
+      float paraB = 2 * (x_init * px + y_init * py) / a;
+      float paraC = 2 * (y_init * px - x_init * py) / a + 2 * (px * px + py * py) / (a * a);
+      float A = paraB * paraB + paraC * paraC;
+      float B = 2 * paraA * paraB;
+      float C = paraA * paraA - paraC * paraC;
+      float sol1 = (-B + alpaka::math::sqrt(acc, B * B - 4 * A * C)) / (2 * A);
+      float sol2 = (-B - alpaka::math::sqrt(acc, B * B - 4 * A * C)) / (2 * A);
+      float solz1 = alpaka::math::asin(acc, sol1) / rou * pz / p + z_init;
+      float solz2 = alpaka::math::asin(acc, sol2) / rou * pz / p + z_init;
+      float diffz1 = (solz1 - z3) * 100;
+      float diffz2 = (solz2 - z3) * 100;
+      if (alpaka::math::isnan(acc, diffz1))
+        residual = diffz2;
+      else if (alpaka::math::isnan(acc, diffz2))
+        residual = diffz1;
+      else {
+        residual = (alpaka::math::abs(acc, diffz1) < alpaka::math::abs(acc, diffz2)) ? diffz1 : diffz2;
+      }
+    } else { // for endcap
+      float s = (z_target - z_init) * p / pz;
+      float x = x_init + px / a * alpaka::math::sin(acc, rou * s) - py / a * (1 - alpaka::math::cos(acc, rou * s));
+      float y = y_init + py / a * alpaka::math::sin(acc, rou * s) + px / a * (1 - alpaka::math::cos(acc, rou * s));
+      residual = (r_target - alpaka::math::sqrt(acc, x * x + y * y)) * 100; 
+    }
+    
+    // error
+    if (moduleType3 == 0) { 
+      error = 0.15f;        //PS
+    } else  {              
+      error = 5.0f;         //2S
+    }
+
+    float projection_missing = 1;
+    if (drdz < 1)
+      projection_missing = ((subdets == lst::Endcap) or (side == lst::Center))
+                                ? 1.f
+                                : 1 / alpaka::math::sqrt(acc, 1 + drdz * drdz);  // cos(atan(drdz)), if dr/dz<1
+    if (drdz > 1)
+      projection_missing = ((subdets == lst::Endcap) or (side == lst::Center))
+                                ? 1.f
+                                : drdz / alpaka::math::sqrt(acc, 1 + drdz * drdz);  //sin(atan(drdz)), if dr/dz>1
```

Comment:
this value is not used by itself, only as a square, replace with `projection_missing2` to also get rid of `sqrt`

Reply:
updated